### ISSUE TITLE
[Feature Request] Support of additional covariance structures (ARH(1) and TOEP)

### ISSUE THREAD

**Issue by pzhou-cims** [glmmTMB/glmmTMB#1095]
Hi, thank you for the incredible work on developing glmmTMB, it's an invaluable tool for many of us working with mixed models!

I'm currently working on a project aimed at replicating the SAS PROC GLIMMIX procedure in R using glmmTMB for categorical MMRM (e.g. where the outcome variable is binomial). While glmmTMB supports several covariance structures, I have encountered a limitation: specifically, ARH(1) and TOEP (Toeplitz, homogenous) are not currently available, but are essential and commonly used for the MMRM analysis that I am trying to replicate.

I’m wondering if it might be possible to include these covariance structures in a future release of glmmTMB, or if you could provide suggestions on how we might implement them?

Thank you in advance for your help.

**Comment by bbolker** [comment]
Neither of these sound too hard.  There are some brief pointers on how to add covariance structures to `glmmTMB` in the ["Hacking glmmTMB" vignette](https://cran.r-project.org/web/packages/glmmTMB/vignettes/hacking.html) (In the "Adding a covariance structure") section.

In the case of homogeneous Toeplitz, this can already be done (as suggested in the "Covariance structures" vignette):

> Homogenous versions of some structures (e.g. Toeplitz, compound symmetric) can be implemented by using the `map` argument to set all log-SD parameters equal to each other.

That is, if you were (for example) fitting a model with a single (heterogeneous) Toeplitz-structured RE with `n` levels, you would specify `map = list(theta = factor(c(rep(1, n), 2:n)))` (I think).  (That said, it would be more convenient to implement a new covariance structure and shouldn't be terribly difficult ... I guess we would call it `homtoep` for consistency with `homdiag` ?

It should be *relatively* easy to implement these because they represent fairly minor modifications of existing structures.

The fastest way to get this done (if you feel up to it) would be to implement these yourself in a GitHub fork and then send a pull request. Otherwise we can add it to the wish list, but it's a long list ...

**Comment by pzhou-cims** [comment]
Thank you for the suggestions.

We will try the map() argument for the homogeneous Toeplitz as you recommended. However, for adding other structures like ARH1, as described in the ["Hacking glmmTMB" vignette](https://cran.r-project.org/web/packages/glmmTMB/vignettes/hacking.html), this would require modification to the C++ code, which is beyond our current capabilities.

That said, we appreciate your quick response on this issue, and it would great if this could be added to the wishlist for a future standard implementation. Thank you!

**Comment by bbolker** [comment]
I did a preliminary implementation of heterogeneous AR1 [here](https://github.com/glmmTMB/glmmTMB/tree/hetar1).  It is **not tested at all** (yet) ... but I have at least nominally gone through all the steps in the "hacking vignette" (except adding documentation), so it *might* work ...
[

**Comment by SchmidtPaul** [comment]
Hi I would simply like to chime in and say that I am also using glmmTMB() a lot but since people in my area grew up with SAS I am still longing for `ARH(1)` as well as Factor-Analytic `FA(q)` - see [SAS documentation](https://documentation.sas.com/doc/en/statug/15.2/statug_mixed_syntax14.htm#statug.mixed.mixedcovstruct)

**Comment by bbolker** [comment]
I think `FA(q)` is approximately equivalent to `rr(..., d=q)` ?

**Comment by bbolker** [comment]
@pzhou-cims  would you be willing to take this for a spin? (`remotes::install_github("glmmTMB/glmmTMB/glmmTMB", ref = "hetar1")`)

**Comment by pzhou-cims** [comment]
> @pzhou-cims would you be willing to take this for a spin? (`remotes::install_github("glmmTMB/glmmTMB/glmmTMB", ref = "hetar1")`)

Thank you so much for the update! We tried implementing arh1 as instructed, however when fitting the model, we encountered the following error (which also appeared for other covariance structures):
![f34aa903-9b5d-4ca4-af5d-8a1600d2c6ad](https://github.com/user-attachments/assets/eb220cf4-9026-4bd6-a6fa-ddb0ae94b00d)

**Comment by bbolker** [comment]
I think this might have been a mismatch based on some changes elsewhere in the code base.  **But** still not working, will take another look ...

**Comment by bbolker** [comment]
FWIW it should be working better than it was before (see newest version of the PR); it's not working perfectly, though (e.g. the printing of results is still messed up).

**Comment by danielinteractive** [comment]
@bbolker I could look into adding the homogeneous Toeplitz structure - so we would call it `homtoep` right?

**Comment by bbolker** [comment]
yes. It should be pretty easy (?), following the "adding a covariance structure" details in the [vignette](https://cran.r-project.org/web/packages/glmmTMB/vignettes/hacking.html) ...

**Comment by danielinteractive** [comment]
Great, thanks @bbolker, great to have the vignette checklist! Here slightly modified by adding documentation todos:

C++ code:

- [x] add to the `valid_covStruct` enum in `glmmTMB.cpp`
- [x] modify `termwise_nll`

R code:

- [x] modify `parFun` to specify the number of parameters
- [x] run `make enum-update`

Documentation:

- [x] modify documentation of `glmmTMB()`
- [x] Add to the covarince structures vignette
- [x] `NEWS` update

Tests:

- [x] Add to `tests/testthat/test-varstruc.R`
